rep-2007.rst

```diff
@@ -0,0 +1,252 @@
+REP: 2007
+Title: Adapting C++ Types
+Author: Audrow Nash <audrow@openrobotics.org>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 28-Jan-2020
+Post-History: 28-Jan-2020
+
+
+Abstract
+========
+
+  A short (~200 word) description of the technical issue being addressed.
+
+This REP proposes an extension to publishers and subscribers in `rclcpp` to make it easier to convert between ROS 2 interfaces and custom C++ data structures.
+
+
+Motivation
+==========
+
+  The motivation is critical for REPs that want to change the ROS APIs. It should clearly explain why the existing API specification is inadequate to address the problem that the REP solves. REP submissions without sufficient motivation may be rejected outright.
+
+The primary reason for this change is to improve the developer's experience working with ROS 2.
```

Comment:
But also, performance. This interface will allow us to define methods for serializing directly to the user requested type, and/or using that type in intra-process communication without ever converting it. Whereas without this feature, even to send a `cv::Mat` from one publisher to a subscription in the same process would require converting it to a `sensor_msgs::msg::Image` first and then back again to `cv::Mat`.